experimental/OrthogonalDiscriminants/src/utils.jl

```diff
@@ -98,6 +100,303 @@ function reduce_mod_squares(val::AbsSimpleNumFieldElem)
+          else
+            # Try to recurse.
+            pi = possible_permutation_characters_from_sylow_subgroup(s, p)
+            if pi != nothing
+              pi = collect(Set([x^tbl for x in pi]))
+              if candlist == nothing
+                candlist = pi
+              else
+                intersect!(candlist, pi)
+              end
+              length(candlist) == 1 && return candlist
+            end
+          end
+        elseif mod(div(order(tbl), order(s)), p) == 0 &&
+               is_prime_power_with_data(div(order(tbl), order(s)))[1]
+          # Compute the perm. char. of the Sylow p-subgroup in the subgroup,
+          # and then try to extend the character uniquely to G.
+          cand = possible_permutation_characters_from_sylow_subgroup(s, p)
+          if cand != nothing
+            extcand = []
+            for pi in cand
+              pi = GAP.Globals.CompositionMaps(pi.values,
+                       GAP.Globals.InverseMap(GapObj(fus)))
+              union!(extcand, [Oscar.class_function(tbl, x) for x in
+                                GAP.Globals.PermChars(Oscar.GAPTable(tbl),
+                                  GAP.GapObj(Dict(:torso => pi), true))])
+            end
+            if candlist == nothing
+              candlist = extcand
+            else
+              intersect!(candlist, extcand)
+            end
+            length(candlist) == 1 && return candlist
+          end
+        end
+      end
+    end
+  end
+
+  isdefined(tbl, :group) && return [trivial_character(sylow_subgroup(group(tbl), p)[1])^tbl]
+
+  # If we arrive here,
+  # we can try to recurse to subgroups for which fusions must be composed,
+  # or to compute candidates.
+  # (Eventually a function that returns all candidates might be useful then.)
+  return candlist
+end
+
+
+# helper function:
+# array of identifiers of available character tables of (non-dashed)
+# automorphic extensions of a simple character table.
+function _identifiers_of_almost_simple_tables(simplename::String)
+  gapname = GapObj(simplename)
+  nams = GAP.Globals.CTblLib.DisplayAtlasMap_ComputePortions(
+           gapname, 0).identifiers[1]
+  return filter(x -> x == gapname || x[end] != '\'', Vector{String}(nams))
+end
+
+# helper function:
+# Let `M` be a vector of vectors representing a *regular* matrix
+# (this is not checked),
+# and let `column_orbits` be the set of orbits of some group `G`
+# of matrix automorphisms of `M` on the positions of columns.
+# Return the set of orbits of the corresponding action of `G`
+# on the rows of `M`.
+function _row_orbits_from_column_orbits(M::Vector, column_orbits::Vector{Vector{Int}})
+  # Compute the auxiliary matrix of orbit sums.
+  n = length(M)
+  MM = [[] for _ in 1:n]
+  for omega in column_orbits
+    for i in 1:n
+      push!(MM[i], sum(M[i][omega]))
+    end
+  end
+
+  # Compute the orbits on rows.
+  row_orbits = Vector{Int}[]
+  reps = []
+  poss = []
```

Comment:
whenever I see `[]` I am think "not type stable, beware". We see a bunch of these in this PR. But perhaps it is fine for other reasons -- just wanted to draw your attention to it.